File: InfiniteAsyncSource.GRPC/IssuesData.Server/IssuesDataProvider.cs
using IssuesData;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace IssuesServer {
    public static class IssuesDataProvider {
        #region helpers
        static object SyncObject = new object();
        static Lazy<Issue[]> AllIssues = new Lazy<Issue[]>(() => {
            var date = DateTime.Today;
            var rnd = new Random(0);
            return Enumerable.Range(0, 100000)
                .Select(i => {
                    date = date.AddSeconds(-rnd.Next(20 * 60));
                    return new Issue { 
                        Id = i,
                        Subject = OutlookDataGenerator.GetSubject(),
                        UserId = OutlookDataGenerator.GetFromId(),
                        Created = date,
                        Votes = rnd.Next(100),
                        Priority = OutlookDataGenerator.GetPriority()
                    };
                }).ToArray();
        });
        class DefaultSortComparer : IComparer<Issue> {
            int IComparer<Issue>.Compare(Issue x, Issue y) {
                if(x.Created.Date != y.Created.Date)
                    return Comparer<DateTime>.Default.Compare(x.Created.Date, y.Created.Date);
                return Comparer<int>.Default.Compare(x.Votes, y.Votes);
            }
        } 
        #endregion

        public static Task<Issue[]> GetIssuesAsync(int skip, int take, IssueSortOrder sortOrder, IssueFilter filter) {
            return Task.Run(() => {
                var issues = SortIssues(sortOrder, AllIssues.Value);
                if(filter != null)
                  issues = FilterIssues(filter, issues);
                return issues.Skip(skip).Take(take).Select(x => x.Clone()).ToArray();
            });
        }

        public static Task<User[]> GetUsersAsync() {
            return Task.Run(() => {
                return OutlookDataGenerator.Users
                    .Select((x, i) => {
                      var split = x.Split(' ');
                      return new User { Id = i, FirstName = split[0], LastName = split[1] };
                    })
                    .ToArray();
            });
        }

        public static Task<IssueSummaries> GetSummariesAsync(IssueFilter filter) {
            return Task.Run(() => {
                var issues = (IEnumerable<Issue>)AllIssues.Value;
                if(filter != null)
                    issues = FilterIssues(filter, issues);
                var lastCreated = issues.Any() ? issues.Max(x => x.Created) : default(DateTime?);
                return new IssueSummaries { Count = issues.Count(), LastCreated = lastCreated };
            });
        }

        public static Task UpdateRowAsync(Issue row) {
            return Task.Run(() => {
                if(row == null)
                    return;
                Issue data = AllIssues.Value.FirstOrDefault(x => x.Id == row.Id);
                if(data == null)
                    return;
                data.Priority = row.Priority;
                data.Subject = row.Subject;
                data.UserId = row.UserId;
                data.Votes = row.Votes;
                data.Created = row.Created;
            });
        }

        #region filter
        static IEnumerable<Issue> FilterIssues(IssueFilter filter, IEnumerable<Issue> issues) {
            if(filter.CreatedFrom != null || filter.CreatedTo != null) {
                if(filter.CreatedFrom == null || filter.CreatedTo == null) {
                    throw new InvalidOperationException();
                }
                issues = issues.Where(x => x.Created >= filter.CreatedFrom.Value && x.Created < filter.CreatedTo);
            }
            if(filter.MinVotes != null) {
                issues = issues.Where(x => x.Votes >= filter.MinVotes.Value);
            }
            if(filter.Priority != null) {
                issues = issues.Where(x => x.Priority == filter.Priority);
            }
            return issues;
        }
        #endregion

        #region sort
        static IEnumerable<Issue> SortIssues(IssueSortOrder sortOrder, IEnumerable<Issue> issues) {
            switch(sortOrder) {
            case IssueSortOrder.Default:
                return issues;//.OrderByDescending(x => x, new DefaultSortComparer()).ThenByDescending(x => x.Created);
            case IssueSortOrder.CreatedDescending:
                return issues.OrderByDescending(x => x.Created);
            case IssueSortOrder.VotesAscending:
                return issues.OrderBy(x => x.Votes).ThenByDescending(x => x.Created);
            case IssueSortOrder.VotesDescending:
                return issues.OrderByDescending(x => x.Votes).ThenByDescending(x => x.Created);
            default:
                throw new InvalidOperationException();
            }
        } 
        #endregion
    }
}

File: VB/IssuesData.Server/IssuesDataProvider.vb
Imports IssuesData
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading.Tasks

Namespace IssuesServer
	Public Module IssuesDataProvider
		#Region "helpers"
		Private SyncObject As New Object()
		Private AllIssues As New Lazy(Of Issue())(Function()
			Dim [date] = DateTime.Today
			Dim rnd = New Random(0)
			Return Enumerable.Range(0, 100000).Select(Function(i)
				[date] = [date].AddSeconds(-rnd.Next(20 * 60))
				Return New Issue With {
					.Id = i,
					.Subject = OutlookDataGenerator.GetSubject(),
					.UserId = OutlookDataGenerator.GetFromId(),
					.Created = [date],
					.Votes = rnd.Next(100),
					.Priority = OutlookDataGenerator.GetPriority()
				}
			End Function).ToArray()
		End Function)
		Private Class DefaultSortComparer
			Implements IComparer(Of Issue)

			Private Function IComparerGeneric_Compare(ByVal x As Issue, ByVal y As Issue) As Integer Implements IComparer(Of Issue).Compare
				If x.Created.Date <> y.Created.Date Then
					Return Comparer(Of DateTime).Default.Compare(x.Created.Date, y.Created.Date)
				End If
				Return Comparer(Of Integer).Default.Compare(x.Votes, y.Votes)
			End Function
		End Class
		#End Region

		Public Function GetIssuesAsync(ByVal skip As Integer, ByVal take As Integer, ByVal sortOrder As IssueSortOrder, ByVal filter As IssueFilter) As Task(Of Issue())
			Return Task.Run(Function()
				Dim issues = SortIssues(sortOrder, AllIssues.Value)
				If filter IsNot Nothing Then
					issues = FilterIssues(filter, issues)
				End If
				Return issues.Skip(skip).Take(take).Select(Function(x) x.Clone()).ToArray()
			End Function)
		End Function

		Public Function GetUsersAsync() As Task(Of User())
			Return Task.Run(Function()
				Return OutlookDataGenerator.Users.Select(Function(x, i)
					Dim split = x.Split(" "c)
					Return New User With {
						.Id = i,
						.FirstName = split(0),
						.LastName = split(1)
					}
				End Function).ToArray()
			End Function)
		End Function

		Public Function GetSummariesAsync(ByVal filter As IssueFilter) As Task(Of IssueSummaries)
			Return Task.Run(Function()
				Dim issues = DirectCast(AllIssues.Value, IEnumerable(Of Issue))
				If filter IsNot Nothing Then
					issues = FilterIssues(filter, issues)
				End If
				Dim lastCreated = If(issues.Any(), issues.Max(Function(x) x.Created), CType(Nothing, DateTime?))
				Return New IssueSummaries With {
					.Count = issues.Count(),
					.LastCreated = lastCreated
				}
			End Function)
		End Function

		Public Function UpdateRowAsync(ByVal row As Issue) As Task
			Return Task.Run(Sub()
				If row Is Nothing Then
					Return
				End If
				Dim data As Issue = AllIssues.Value.FirstOrDefault(Function(x) x.Id = row.Id)
				If data Is Nothing Then
					Return
				End If
				data.Priority = row.Priority
				data.Subject = row.Subject
				data.UserId = row.UserId
				data.Votes = row.Votes
				data.Created = row.Created
			End Sub)
		End Function

		#Region "filter"
		Private Function FilterIssues(ByVal filter As IssueFilter, ByVal issues As IEnumerable(Of Issue)) As IEnumerable(Of Issue)
			If filter.CreatedFrom IsNot Nothing OrElse filter.CreatedTo IsNot Nothing Then
				If filter.CreatedFrom Is Nothing OrElse filter.CreatedTo Is Nothing Then
					Throw New InvalidOperationException()
				End If
'INSTANT VB WARNING: Comparisons involving nullable type instances require Option Strict Off:
				issues = issues.Where(Function(x) x.Created >= filter.CreatedFrom.Value AndAlso x.Created < filter.CreatedTo)
			End If
			If filter.MinVotes IsNot Nothing Then
				issues = issues.Where(Function(x) x.Votes >= filter.MinVotes.Value)
			End If
			If filter.Priority IsNot Nothing Then
				issues = issues.Where(Function(x) filter.Priority.Equals(x.Priority))
			End If
			Return issues
		End Function
		#End Region

		#Region "sort"
		Private Function SortIssues(ByVal sortOrder As IssueSortOrder, ByVal issues As IEnumerable(Of Issue)) As IEnumerable(Of Issue)
			Select Case sortOrder
			Case IssueSortOrder.Default
				Return issues '.OrderByDescending(x => x, new DefaultSortComparer()).ThenByDescending(x => x.Created);
			Case IssueSortOrder.CreatedDescending
				Return issues.OrderByDescending(Function(x) x.Created)
			Case IssueSortOrder.VotesAscending
				Return issues.OrderBy(Function(x) x.Votes).ThenByDescending(Function(x) x.Created)
			Case IssueSortOrder.VotesDescending
				Return issues.OrderByDescending(Function(x) x.Votes).ThenByDescending(Function(x) x.Created)
			Case Else
				Throw New InvalidOperationException()
			End Select
		End Function
		#End Region
	End Module
End Namespace
